### PR TITLE
More pigstensions

### PR DESCRIPTION
**Pull request by sfilipi**
ongoing work to address #754. 
This PR adds the xtensions for AP, OGD, LR, Multi-LR, and Poisson Regression.

### CODE REVIEW

File: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsStatics.cs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        /// <param name="l2Weight">Weight of L2 regularizer term.</param>
+        /// <param name="memorySize">Memory size for <see cref="Runtime.Learners.LogisticRegression"/>. Lower=faster, less accurate.</param>
+        /// <param name="optimizationTolerance">Threshold for optimizer convergence.</param>
+        /// <param name="onFit">A delegate that is called every time the
+        /// <see cref="Estimator{TTupleInShape, TTupleOutShape, TTransformer}.Fit(DataView{TTupleInShape})"/> method is called on the
+        /// <see cref="Estimator{TTupleInShape, TTupleOutShape, TTransformer}"/> instance created out of this. This delegate will receive
+        /// the linear model that was trained.  Note that this action cannot change the result in any way; it is only a way for the caller to
+        /// be informed about what was learnt.</param>
+        /// <returns>The predicted output.</returns>
+        public static (Scalar<float> score, Scalar<float> probability, Scalar<bool> predictedLabel) LogisticRegression(this BinaryClassificationContext.BinaryClassificationTrainers ctx,
```
Comment (sfilipi) [inline src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsStatics.cs]: >LogisticRegression [](start = 100, length = 18)

rename to LogisticRegressionBinaryClassifier

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLearnerStatic.cs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+                            var model = trans.Model;
+                            if (model is ParameterMixingCalibratedPredictor cali)
+                                onFit((LinearBinaryPredictor)cali.SubPredictor);
+                            else
+                                onFit((LinearBinaryPredictor)model);
+                        });
+                    }
+                    return trainer;
+                    */
+
```
Comment (sfilipi) [inline src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLearnerStatic.cs]: remove

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLearnerStatic.cs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+using System;
+
+namespace Microsoft.ML.Trainers
+{
+    public static class OnlineLearnerStatic
+    {
+        /// <summary>
+        /// Predict a target using a linear binary classification model trained with the SDCA trainer, and a custom loss.
+        /// Note that because we cannot be sure that all loss functions will produce naturally calibrated outputs, setting
+        /// a custom loss function will not produce a calibrated probability column.
```
Comment (sfilipi) [inline src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLearnerStatic.cs]: remove

File: test/Microsoft.ML.StaticPipelineTesting/Training.cs [hunk 4]
```diff
@@ -160,9 +160,9 @@ public void SdcaBinaryClassificationNoClaibration()
 
             // With a custom loss function we no longer get calibrated predictions.
             var est = reader.MakeNewEstimator()
-                .Append(r => (r.label, preds: ctx.Trainers.Sdca(r.label, r.features,
-                maxIterations: 2,
-                loss: loss, onFit: p => pred = p)));
+                .Append(r => (r.label, preds: ctx.Trainers.AveragedPerceptron(loss, r.label, r.features,
```
Comment (sfilipi) [inline test/Microsoft.ML.StaticPipelineTesting/Training.cs]: >AveragedPerceptron [](start = 59, length = 18)

revert

File: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsStatics.cs [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+using System;
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.Internal.Calibration;
+using Microsoft.ML.Runtime.Learners;
+using Microsoft.ML.Runtime.Training;
+using Microsoft.ML.StaticPipe;
+using Microsoft.ML.StaticPipe.Runtime;
+
+namespace Microsoft.ML.Trainers
```
Comment (sfilipi) [inline src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsStatics.cs]: >Microsoft.ML.Trainers [](start = 10, length = 21)

change namespace

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLearnerStatic.cs [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.Internal.Calibration;
+using Microsoft.ML.Runtime.Learners;
+using Microsoft.ML.StaticPipe;
+using Microsoft.ML.StaticPipe.Runtime;
+using System;
+
+namespace Microsoft.ML.Trainers
```
Comment (sfilipi) [inline src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLearnerStatic.cs]: >Microsoft.ML.Trainers [](start = 10, length = 21)

change namespace

File: test/Microsoft.ML.Tests/TrainerEstimators/OnlineLinearTests.cs [hunk 7]
```diff
@@ -25,7 +25,7 @@ public void OnlineLinearWorkout()
 
             var trainData = pipe.Fit(data).Transform(data).AsDynamic;
 
-            IEstimator<ITransformer> est = new OnlineGradientDescentTrainer(Env, new OnlineGradientDescentTrainer.Arguments());
+            IEstimator<ITransformer> est = new OnlineGradientDescentTrainer(Env, "Label", "Features");
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Tests/TrainerEstimators/OnlineLinearTests.cs]: >"Label" [](start = 81, length = 7)

Any chance I can convince you to use DefaultColumnNames.Label?

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs [hunk 8]
```diff
@@ -46,27 +46,69 @@ public sealed class Arguments : AveragedLinearArguments
...
+        /// <param name="labelColumn">Name of the label column.</param>
+        /// <param name="featureColumn">Name of the feature column.</param>
+        /// <param name="learningRate">The learning Rate.</param>
+        /// <param name="decreaseLearningRate">Decrease learning rate as iterations progress.</param>
+        /// <param name="l2RegularizerWeight">L2 Regularization Weight.</param>
+        /// <param name="numIterations">Number of training iterations through the data.</param>
+        /// <param name="weightsColumn">The name of the weights column.</param>
+        /// <param name="lossFunction">The custom loss functions. Defaults to <see cref="SquaredLoss"/> if not provided.</param>
+        public OnlineGradientDescentTrainer(IHostEnvironment env,
+            string labelColumn,
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs]: >labelColumn [](start = 19, length = 11)

any reason not to use default values here?

File: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs [hunk 9]
```diff
@@ -164,18 +184,24 @@ internal LbfgsTrainerBase(IHostEnvironment env, TArgs args, SchemaShape.Column l
             Contracts.CheckUserArg(Args.SgdInitializationTolerance >= 0, nameof(Args.SgdInitializationTolerance), "Must be non-negative");
             Contracts.CheckUserArg(Args.NumThreads == null || Args.NumThreads.Value >= 0, nameof(Args.NumThreads), "Must be non-negative");
 
-            L2Weight = Args.L2Weight;
-            L1Weight = Args.L1Weight;
-            OptTol = Args.OptTol;
-            MemorySize = Args.MemorySize;
+            Contracts.CheckParam(l2Weight >= 0, nameof(l2Weight), "Must be non-negative, if provided.");
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs]: >Contracts [](start = 12, length = 9)

why not Host?

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLearnerStatic.cs [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+                IClassificationLoss lossFunction,
+                Scalar<bool> label, Vector<float> features, Scalar<float> weights = null,
+                float learningRate = AveragedLinearArguments.AveragedDefaultArgs.LearningRate,
+                bool decreaseLearningRate = AveragedLinearArguments.AveragedDefaultArgs.DecreaseLearningRate,
+                float l2RegularizerWeight = AveragedLinearArguments.AveragedDefaultArgs.L2RegularizerWeight,
+                int numIterations = AveragedLinearArguments.AveragedDefaultArgs.NumIterations,
+                Action<LinearBinaryPredictor> onFit = null
+            )
+        {
+            Contracts.CheckValue(label, nameof(label));
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLearnerStatic.cs]: >Contracts [](start = 12, length = 9)

technically you have ctx.Environment

Comment (sfilipi) [reply src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLearnerStatic.cs]: ctx is just the trainers, not the context. I don't have the env.

---
In reply to: [221115177](https://github.com/dotnet/machinelearning/pull/1084#discussion_r221115177) [](ancestors = 221115177)

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs (hunk 8 above)
Comment (sfilipi) [reply src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs]: i think we have marked the label and feature columns as required everywhere. 

---
In reply to: [221113276](https://github.com/dotnet/machinelearning/pull/1084#discussion_r221113276) [](ancestors = 221113276)

File: test/Microsoft.ML.Tests/TrainerEstimators/OnlineLinearTests.cs (hunk 7 above)
Comment (sfilipi) [reply test/Microsoft.ML.Tests/TrainerEstimators/OnlineLinearTests.cs]: I can change, it; it won't be any better than it is. 
This is "Label", because in like 20 we defined r.Label. 

There is no way to type the the member name to be the same as a particular string.
(might be the only thing i miss from javascript :P)

---
In reply to: [221112970](https://github.com/dotnet/machinelearning/pull/1084#discussion_r221112970) [](ancestors = 221112970)

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLearnerStatic.cs (hunk 6 above)
Comment (sfilipi) [reply src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLearnerStatic.cs]: Actually moving to ML.StaticPipe namespace won't fix it by itself. All the contexts are in Microsoft.ML. 

@Tomfinley move the contexts in StaticPipe as well?

---
In reply to: [221111449](https://github.com/dotnet/machinelearning/pull/1084#discussion_r221111449) [](ancestors = 221111449)

File: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsStatics.cs (hunk 5 above)
Comment (sfilipi) [reply src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsStatics.cs]: discussion on the other comment. 

---
In reply to: [221111414](https://github.com/dotnet/machinelearning/pull/1084#discussion_r221111414) [](ancestors = 221111414)

File: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs (hunk 9 above)
Comment (sfilipi) [reply src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs]: thanks!

---
In reply to: [221114697](https://github.com/dotnet/machinelearning/pull/1084#discussion_r221114697) [](ancestors = 221114697)

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLearnerStatic.cs (hunk 6 above)
Comment (Zruty0) [reply src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLearnerStatic.cs]: Yes

---
In reply to: [221299214](https://github.com/dotnet/machinelearning/pull/1084#discussion_r221299214) [](ancestors = 221299214,221111449)

Comment (Zruty0) [reply src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLearnerStatic.cs]: No. If the user opts into the static pipe, they will see the trainers. It's good enough for me

---
In reply to: [221299214](https://github.com/dotnet/machinelearning/pull/1084#discussion_r221299214) [](ancestors = 221299214,221111449)

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLearnerStatic.cs (hunk 10 above)
Comment (Zruty0) [reply src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLearnerStatic.cs]: Oh, you do. You need to call `StaticPipeUtils.GetEnvironment` to get it

---
In reply to: [221295921](https://github.com/dotnet/machinelearning/pull/1084#discussion_r221295921) [](ancestors = 221295921,221115177)

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs (hunk 8 above)
Comment (Zruty0) [reply src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs]: I think we were not consistent. 

---
In reply to: [221298419](https://github.com/dotnet/machinelearning/pull/1084#discussion_r221298419) [](ancestors = 221298419,221113276)

File: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsStatics.cs [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+                        return trainer.WithOnFitDelegate(trans => onFit(trans.Model));
+                    return trainer;
+                }, label, features, weights);
+
+            return rec.Output;
+        }
+
+    }
+
+    internal sealed class LbfgsStaticsUtils{
```
Comment (Zruty0) [inline src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsStatics.cs]: >Statics [](start = 31, length = 7)

`static` ?

Comment (Zruty0) [reply src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsStatics.cs]: also the class itself should be `static`

---
In reply to: [221307856](https://github.com/dotnet/machinelearning/pull/1084#discussion_r221307856) [](ancestors = 221307856)

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLearnerStatic.cs (hunk 6 above)
Comment (sfilipi) [reply src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLearnerStatic.cs]: The context is in Microsoft.ML, i don't see anything implicitely import staticPipe.

---
In reply to: [221307154](https://github.com/dotnet/machinelearning/pull/1084#discussion_r221307154) [](ancestors = 221307154,221299214,221111449)

File: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsStatics.cs [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+    {
+        /// <summary>
+        ///  Predict a target using a linear binary classification model trained with the <see cref="Runtime.Learners.LogisticRegression"/> trainer.
+        /// </summary>
+        /// <param name="ctx">The binary classificaiton context trainer object.</param>
+        /// <param name="label">The label, or dependent variable.</param>
+        /// <param name="features">The features, or independent variables.</param>
+        /// <param name="weights">The optional example weights.</param>
+        /// <param name="enoforceNoNegativity">Enforce non-negative weights.</param>
+        /// <param name="l1Weight">Weight of L1 regularizer term.</param>
```
Comment (artidoro) [inline src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsStatics.cs]: Maybe it is more standard to say regularization, instead of regularizer.

**Comment by artidoro** [comment]
> using Float = System.Single;

If you have time, would be a good idea to clean this up.

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs:5 in f414717. [](commit_id = f41471707f263a213f26360bdf33ba29008ea995, deletion_comment = False)

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs [hunk 13]
```diff
@@ -46,27 +46,69 @@ public sealed class Arguments : AveragedLinearArguments
...
+            string featureColumn,
+            float learningRate = Arguments.OgdDefaultArgs.LearningRate,
+            bool decreaseLearningRate =Arguments.OgdDefaultArgs.DecreaseLearningRate,
```
Comment (artidoro) [inline src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs]: Add a space after equal.

File: src/Microsoft.ML.StandardLearners/Standard/PoissonRegression/PoissonRegression.cs [hunk 14]
```diff
@@ -46,10 +46,22 @@ public sealed class Arguments : ArgumentsBase
         /// <param name="labelColumn">The name of the label column.</param>
         /// <param name="featureColumn">The name of the feature column.</param>
         /// <param name="weightColumn">The name for the example weight column.</param>
+        /// <param name="enforceNoNegativity">Enforce non-negative weights.</param>
+        /// <param name="l1Weight">Weight of L1 regularizer term.</param>
```
Comment (artidoro) [inline src/Microsoft.ML.StandardLearners/Standard/PoissonRegression/PoissonRegression.cs]: >regularizer [](start = 48, length = 11)

regularizer pops up in many files, let me see if I can help you find them.
